### PR TITLE
[MRG] Add SPDNet

### PR DESCRIPTION
**Pull request by tgnassou**
Hi, I started implementing SPDNet in Braindecode, as discussed in the issue [#453](https://github.com/braindecode/braindecode/issues/453).  
It is just a draft, and the model is not working. I added an example if someone wants to try it. I still need the pooling and batch normalization layers to have the same SPDNet as [Huang et al.](https://arxiv.org/abs/1608.04233). I will also add the Riemannian batch norm of [Kobler et al.](https://arxiv.org/abs/2206.01323) and [Brooks et al.](https://arxiv.org/abs/1909.02414). 

I am new in this area of SPDNet. If anyone has any insights or suggestions, please let me know.

### CODE REVIEW

**Comment by agramfort** [comment]
Ping @rkobler

**Comment by bruAristimunha** [comment]
Heyyy @tgnassou!!

I added the requirements to your example compile. I added Pyriemann, because I planned to push a small tutorial on how to mix braindecode and pyriemann related to my research with @sylvchev. I hope you don't mind.

File: setup.py [hunk 1]
```diff
@@ -40,6 +40,7 @@
         'docs': ['sphinx_gallery', 'sphinx_rtd_theme', 'pydata_sphinx_theme', 'numpydoc',
                  'memory_profiler', 'pillow', 'ipython', 'sphinx_design', 'docstring_inheritance',
                  'lightning'],
+        'riemann': ['pyriemann', 'geoopt'],
```
Comment (agramfort) [inline setup.py:43]: if it's just for an example I would put this in docs key no? it becomes required for the doc public.

also how do you test now that braindecode can work without geoopt?

Comment (bruAristimunha) [reply setup.py:43]: hmm, true @agramfort!

Comment (bruAristimunha) [reply setup.py]: solved @agramfort!

**Comment by jpaillard** [comment]
Hi !
Interesting implementation, thx for sharing.
I was wondering if there is a reason why you adopted Riemannian gradients with geoopt instead of a parameterized optimization approach such as [GeoTorch](https://github.com/lezcano/geotorch). 
According to [this paper (theorem 4.3)](https://proceedings.neurips.cc/paper_files/paper/2019/file/1b33d16fc562464579b7199ca3114982-Paper.pdf)  both approaches theoretically lead to the same outcome but maybe there is a numerical difference ? 
If it makes sense I can provide an implementation using GeoTorch for comparison :)

**Comment by rkobler** [comment]
Hi @tgnassou!

Great that you started to implement SPD type models in braindecode. I am happy to support and help in the implementation process!

@jpaillard it would be definitely interesting to test both approaches. So far, I only have experience with geoopt.

**Comment by bruAristimunha** [comment]
The review is yours @agramfort :)

File: braindecode/models/functions.py [hunk 2]
```diff
@@ -45,3 +45,104 @@ def transpose_time_to_spat(x):
...
+
+    @staticmethod
+    def backward(ctx, grad_output):
+        s, U, s_modified = ctx.saved_tensors
+        def derivative(s): return s.reciprocal()
+        return modeig_backward(grad_output, s, U, s_modified, derivative)
+
+
+class regm(torch.autograd.Function):
+    """Regularized matrix logarithm of a symmetric matrix.
```
Comment (agramfort) [inline braindecode/models/functions.py:116]: it's not regularized in the classic sense where you add a constant on the diag. It puts all values less then threshold to 0 so it outputs rank deficient matrices unless I miss something.

File: braindecode/models/spdnet.py [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+    def forward(self, X):
+        return self.W.transpose(-1, -2) @ X @ self.W
+
+
+class ReEig(nn.Module):
+    """ReEig layer from [1]_.
+
+    This class add non-linearity to the network by
+    applying a rectified linear unit to the eigenvalues
+    of a symmetric matrix.
```
Comment (agramfort) [inline braindecode/models/spdnet.py:91]: Say that by doing this it makes the matrix non-negative and positive if threshold > 0.

File: braindecode/models/spdnet.py [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+
+    def forward(self, X):
+        return regm.apply(X, self.threshold_)
+
+
+class LogEig(nn.Module):
+    """LogEig layer from [1]_.
+
+    This class perform Riemannian projection into a flat space
+    by applying the logarithm to the eigenvalues of a symmetric matrix.
```
Comment (agramfort) [inline braindecode/models/spdnet.py:121]: say also that is flattens the output as the logm returns a matrix not a vector in maths/scipy.

File: braindecode/models/spdnet.py [hunk 5]
```diff
@@ -0,0 +1,238 @@
...
+
+class SPDNet(EEGModuleMixin, nn.Module):
+    """SPDNet from [1]_.
+
+    This class is a SPDNet implementation from [1]_.
+
+    Parameters
+    ----------
+    input_type : str
+        Input type
```
Comment (agramfort) [inline braindecode/models/spdnet.py:171]:
```suggestion
```

File: braindecode/models/spdnet.py [hunk 6]
```diff
@@ -0,0 +1,238 @@
...
+           AAAI
+    """
+    def __init__(
+        self,
+        input_type,
+        n_chans,
+        subspacedim,
+        threshold=1e-4,
+        n_outputs=1,
+        chs_info=None,
```
Comment (agramfort) [inline braindecode/models/spdnet.py:200]: check that all parameters are documented in the docstring

File: test/unit_tests/models/test_models.py [hunk 7]
```diff
@@ -657,6 +658,59 @@ def hook(model, input, output):
...
+        input_type="cov",
+        n_chans=n_channels,
+        n_outputs=n_classes,
+        subspacedim=subspacedim,
+        n_times=int(sfreq * input_size_s)
+    )
+    model.eval()
+
+    y_pred2 = model(cov)  # 3D inputs
+    assert y_pred2.shape == (n_examples, n_classes)
```
Comment (agramfort) [inline test/unit_tests/models/test_models.py:711]: here you test that it runs. I would argue to test that the loss goes down as otherwise you don't check the correctness of the code.

File: braindecode/models/functions.py [hunk 8]
```diff
@@ -45,3 +45,104 @@ def transpose_time_to_spat(x):
...
+        output, s, U, s_modified = modeig_forward(X, function_applied)
+        ctx.save_for_backward(s, U, s_modified)
+        ctx.threshold_ = threshold
+        return output
+
+    @staticmethod
+    def backward(ctx, grad_output):
+        s, U, s_modified = ctx.saved_tensors
+        def derivative(s): return s > ctx.threshold_
+        return modeig_backward(grad_output, s, U, s_modified, derivative), None
```
[the remark this answers is not in the document]
Comment (tgnassou) [reply braindecode/models/functions.py:148]: For SPDNet, it makes sense, but for TSMNet, for example, you will need a logm independently of regm. And it's clearer to have LogEig and ReEig layers to fit the paper.

File: braindecode/models/spdnet.py (hunk 6 above)
Comment (tgnassou) [reply braindecode/models/spdnet.py:200]: Only some of the parameters are documented, but it is the same in several models in braindecode. I didn't know if I needed to add them or not. If this is a mistake in other models, a fix doc PR is needed. WDYT @bruAristimunha ?

Comment (agramfort) [reply braindecode/models/spdnet.py:189]: does it mean that here chs_info cannot be set (/ does not make sense) for SPDNet?

Comment (bruAristimunha) [reply braindecode/models/spdnet.py:189]: Pinging @PierreGtch

File: braindecode/models/spdnet.py [hunk 9]
```diff
@@ -0,0 +1,227 @@
...
+
+    Parameters
+    ----------
+    input_type : str
+        If "raw", the input is a batch of raw EEG signals and
+        a covariance matrix is computed
+        If "cov", the input is a batch of covariance matrices and
+        the covariance matrix is used as input
+    n_chans : int
+        Number of channels
```
Comment (PierreGtch) [inline braindecode/models/spdnet.py:166]:
```suggestion
```
no need, it will automatically be added by `EEGModuleMixin`

File: braindecode/models/spdnet.py [hunk 10]
```diff
@@ -0,0 +1,227 @@
...
+    subspacedim : int
+        Subspace dimension
+    threshold : float
+        Threshold for the rectified linear unit
+    n_outputs : int
+        Output shape
```
Comment (PierreGtch) [inline braindecode/models/spdnet.py:172]:
```suggestion
```
same

File: braindecode/models/spdnet.py [hunk 11]
```diff
@@ -0,0 +1,227 @@
...
+           A Riemannian Network for SPD Matrix Learning
+           AAAI
+    """
+    def __init__(
+        self,
+        input_type,
+        n_chans,
+        subspacedim,
+        threshold=1e-4,
+        n_outputs=1,
```
Comment (PierreGtch) [inline braindecode/models/spdnet.py:188]: Why do you use 1 as default value for `n_outputs` and not `None`?

File: braindecode/models/spdnet.py [hunk 12]
```diff
@@ -0,0 +1,227 @@
...
+        self.bimap = BiMap(n_chans, subspacedim)
+        self.reeig = ReEig(threshold)
+        self.logeig = LogEig(subspacedim, tril=tril)
+        self.len_last_layer = subspacedim * (subspacedim + 1) // 2 if tril else subspacedim**2
+        self.classifier = torch.nn.Linear(self.len_last_layer, n_outputs)
+
+        if add_log_softmax:
+            self.logsoftmax = nn.LogSoftmax(dim=1)
+        else:
+            self.logsoftmax = nn.Identity()
```
Comment (PierreGtch) [inline braindecode/models/spdnet.py:218]:
```suggestion
```
The goal in the long term is to get rid of the log softmax layer in every model (cf #511). The parameter `add_log_softmax` was introduced to throw a depreciation for the models that had one. But new models should not have a log softmax layer at all.

File: braindecode/models/spdnet.py [hunk 13]
```diff
@@ -0,0 +1,227 @@
...
+        tril=True,
+    ):
+        super().__init__(
+            n_outputs=n_outputs,
+            n_chans=n_chans,
+            chs_info=chs_info,
+            n_times=n_times,
+            input_window_seconds=input_window_seconds,
+            sfreq=sfreq,
+            add_log_softmax=add_log_softmax,
```
Comment (PierreGtch) [inline braindecode/models/spdnet.py:203]:
```suggestion
```

File: braindecode/models/spdnet.py [hunk 14]
```diff
@@ -0,0 +1,227 @@
...
+        input_type,
+        n_chans,
+        subspacedim,
+        threshold=1e-4,
+        n_outputs=1,
+        chs_info=None,
+        n_times=None,
+        input_window_seconds=None,
+        sfreq=None,
+        add_log_softmax=False,
```
Comment (PierreGtch) [inline braindecode/models/spdnet.py:193]:
```suggestion
```

File: braindecode/models/spdnet.py [hunk 15]
```diff
@@ -0,0 +1,227 @@
...
+            self.logsoftmax = nn.Identity()
+
+    def forward(self, X):
+        X = self.cov(X)
+        X = self.bimap(X)
+        X = self.reeig(X)
+        X = self.logeig(X)
+        X = self.classifier(X)
+        output = self.logsoftmax(X)
+        return output
```
Comment (PierreGtch) [inline braindecode/models/spdnet.py:227]:
```suggestion
        return X
```

File: braindecode/models/spdnet.py [hunk 16]
```diff
@@ -0,0 +1,227 @@
...
+            n_times=n_times,
+            input_window_seconds=input_window_seconds,
+            sfreq=sfreq,
+            add_log_softmax=add_log_softmax,
+        )
+        if input_type == "raw":
+            self.cov = CovLayer()
+        elif input_type == "cov":
+            self.cov = nn.Identity()
+        self.bimap = BiMap(n_chans, subspacedim)
```
Comment (PierreGtch) [inline braindecode/models/spdnet.py:209]:
```suggestion
        self.bimap = BiMap(self.n_chans, subspacedim)
```
So that `n_chans` can be inferred in case only `chs_info` is passed

File: braindecode/models/spdnet.py [hunk 17]
```diff
@@ -0,0 +1,227 @@
...
+        self.reeig = ReEig(threshold)
+        self.logeig = LogEig(subspacedim, tril=tril)
+        self.len_last_layer = subspacedim * (subspacedim + 1) // 2 if tril else subspacedim**2
+        self.classifier = torch.nn.Linear(self.len_last_layer, n_outputs)
```
Comment (PierreGtch) [inline braindecode/models/spdnet.py:213]:
```suggestion
        self.classifier = torch.nn.Linear(self.len_last_layer, self.n_outputs)
```

File: braindecode/models/spdnet.py (hunk 12 above)
Comment (tgnassou) [reply braindecode/models/spdnet.py:218]: Thanks, I wasn't sure if I needed to put it or not.

**Comment by bruAristimunha** [comment]
Hey @tgnassou and @agramfort, any update about this PR?

**Review by bruAristimunha — COMMENTED** [review]
solving some comments

**Comment by bruAristimunha** [comment]
@tgnassou, can I assume this PR? I think will be super convenient with my paper that I'm finishing this month about SPDNet.

**Comment by tgnassou** [comment]
Yeah sure!

**Comment by bruAristimunha** [comment]
thank you @tgnassou

**Comment by bruAristimunha** [comment]
Things exploded when I updated the branch. I will re-start in a fresh branch/PR @tgnassou
